### PR TITLE
feat(experimental/geo): extend with GIS functions and units support

### PR DESCRIPTION
**Pull request by alespour**
This PR 
- adds several GIS `ST_xxx` functions that appear to be most useful with tracking use cases: `ST_Contains, ST_Distance, ST_DWithin, ST_Intersects, ST_Length`, plus non-standard `ST_LineString`
- adds support for configurable units (as of now, unit for distance can be specified in `km`, `mile` and `m`).
- adds `s2CellLatLon` function that returns coordinates for given s2 Cell (this useful in aggregation queries with `reduce`)
- documents how to implemented geofencing with custom check queries

Includes minor refactorings.

### Done checklist
- [ ] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: stdlib/experimental/geo/README.md [hunk 1]
```diff
@@ -254,14 +292,231 @@ asTracks = (tables=<-, groupBy=["id","tid"], orderBy=["_time"]) =>
...
+
+Example:
+```js
+box = {minLat: 40.51757813, maxLat: 40.86914063, minLon: -73.65234375, maxLon: -72.94921875}
+
+from(bucket:"mta")
+    ...
+    |> geo.toRows()
+    |> map(fn: (r) => ({
+      r with st_within: ST_DWithin(region: region, geometry: {lat: r.lat, lon: r.lon}, distance: 15.0)
```
Comment (wolffcm) [inline stdlib/experimental/geo/README.md:392]: I think you mean `ST_DWithin(region: box, ...)` here, right?

File: stdlib/experimental/geo/README.md [hunk 2]
```diff
@@ -254,14 +292,231 @@ asTracks = (tables=<-, groupBy=["id","tid"], orderBy=["_time"]) =>
...
+    |> reduce(fn: (r, accumulator) => ({
+        r with
+        __linestring: accumulator.__linestring + (if accumulator.__count > 0 then ", " else "") + string(v: r.lat) + " " + string(v: r.lon),
+        __count: accumulator.__count + 1
+      }), identity: {
+        __linestring: "",
+        __count: 0
+      }
+    )
+    |> rename(columns: {__linestring: "st_linestring"})
```
Comment (wolffcm) [inline stdlib/experimental/geo/README.md:477]: Very nice!

File: stdlib/experimental/geo/README.md [hunk 3]
```diff
@@ -254,14 +292,231 @@ asTracks = (tables=<-, groupBy=["id","tid"], orderBy=["_time"]) =>
...
+```js
+box = {minLat: 40.51757813, maxLat: 40.86914063, minLon: -73.65234375, maxLon: -72.94921875}
+
+from(bucket:"mta")
+    ...
+    |> geo.toRows()
+    |> geo.asTracks()
+    |> geo.ST_LineString()
+    |> map(fn: (r) => ({
+      r with st_intersects: ST_Intersects(region: region, geometry: {linestring: r.st_linestring})
```
Comment (wolffcm) [inline stdlib/experimental/geo/README.md:418]: I think you mean `region: box` here too.

File: stdlib/experimental/geo/README.md [hunk 4]
```diff
@@ -254,14 +292,231 @@ asTracks = (tables=<-, groupBy=["id","tid"], orderBy=["_time"]) =>
...
+    |> geo.asTracks()
+    |> geo.ST_LineString()
+    |> map(fn: (r) => ({
+      r with st_length: ST_Length(geometry: {linestring: r.st_linestring})
+    }))
+```
+
+### Function `ST_LineString`
+
+_This is a proprietary function._
```
Comment (wolffcm) [inline stdlib/experimental/geo/README.md:448]: Is this something that some other vendors support but that is not standard, or something totally specific to Flux?

File: libflux/src/core/semantic/builtins.rs [hunk 5]
```diff
@@ -55,10 +55,13 @@ pub fn builtins() -> Builtins<'static> {
                      "from" => "forall [t0] where t0: Row (token: string, project: string, instance: string, table: string) -> [t0]",
             },
             "experimental/geo" => semantic_map! {
-                     "containsLatLon" => "forall [t0] where t0: Row (region: t0, lat: float, lon: float) -> bool",
-                     "getGrid" => "forall [t0] where t0: Row (region: t0, ?minSize: int, ?maxSize: int, ?level: int, ?maxLevel: int) -> {level: int | set: [string]}",
+                     "getGrid" => "forall [t0] where t0: Row (region: t0, ?minSize: int, ?maxSize: int, ?level: int, ?maxLevel: int, units: {distance: string}) -> {level: int | set: [string]}",
                      "getLevel" => "forall [] (token: string) -> int",
```
[the remark this answers is not in the document]
Comment (alespour) [reply libflux/src/core/semantic/builtins.rs:59]: Yes the are the only breaking changes here. Neither of these two functions have been intended to be used directly in user Flux code.

File: stdlib/experimental/geo/README.md (hunk 3 above)
Comment (alespour) [reply stdlib/experimental/geo/README.md:418]: Absolutely. Thank you for such thorough review.

File: stdlib/experimental/geo/README.md (hunk 4 above)
Comment (alespour) [reply stdlib/experimental/geo/README.md:448]: Totally specific to Flux.

It "converts" a series containing points with _lat_,_lon_ to a string formatted as LINESTRING which is a GIS type for representing simple track.

Comment (alespour) [reply stdlib/experimental/geo/README.md]: It is wrong description, such function is actually quite common.

File: stdlib/experimental/geo/README.md (hunk 2 above)
Comment (alespour) [reply stdlib/experimental/geo/README.md:475]: Thank you!

File: stdlib/experimental/geo/geo.go [hunk 6]
```diff
@@ -71,42 +125,15 @@ func generateGetGridFunc() values.Function {
...
-				region = getLoopRegion(points)
+			switch v := geom.(type) {
+			case box:
+				region = getS2RectRegion(v)
+			case circle:
+				region = getS2CapRegion(v)
+			case polygon:
+				region = getS2LoopRegion(v)
+			default:
+				return nil, fmt.Errorf("code %d: unsupported point type: %T", codes.Invalid, geom)
```
Comment (wolffcm) [inline stdlib/experimental/geo/geo.go:136]: Can you to change this to
```go

import "github.com/influxdata/flux/internal/errors"

...

return nil, errors.Newf(codes.Invalid, "unsupported point type: %T", geom)
```

We need to return an instance of this special struct to ensure that we return a 4XX (`codes.Invalid`) or 5XX (`codes.Internal`) from our HTTP API.

Could I trouble you to change the other errors in the `geo` package similarly as well?

Comment (alespour) [reply stdlib/experimental/geo/geo.go]: Done. Thank you.

**Pull request merged**